### ISSUE TITLE
Input Feature Question

### ISSUE THREAD

**Issue by almas2019** [webstah/self-supervised-bss-via-multi-encoder-ae#1]
Hello,

I'm wondering what the input feature dimensions of the ECG and PPG data? Also,  how is the input_length is used in `class ConvolutionalEncoder(nn.Module):`?

**Comment by webstah** [comment]
The input dimensions are defined in the experiment config files here: https://github.com/webstah/self-supervised-bss-via-multi-encoder-ae/tree/main/config/experiment_config

The input length should be 12288 for the ECG and PPG experiments as described in our preprint. The input_length parameter was primarily used for earlier versions of the experiments that had a fully connected layer at the encoding space. This bit of vestigial code can be removed. If you run the model with our recommended configurations it has no effect but may be important in the decoder depending on the layer normalization method used.

I see that the ECG experiment config has the wrong input length set. While this has no impact on the experiment configuration we used, I will update this for clarity's sake.

**Comment by webstah** [comment]
> Thanks, also following up in https://github.com/webstah/self-supervised-bss-via-multi-encoder-ae/blob/main/config/experiment_config/mesa_ecg_bss.yaml , I see there are 7 channels and num_encoders = 8, but in https://github.com/webstah/self-supervised-bss-via-multi-encoder-ae/blob/main/config/experiment_config/mesa_ppg_bss.yaml, there are 8 channels and 8 encoders, is there a reason for this difference or a typo?

Good catch, there is a typo or I accidentally deleted something when copying this file over from my private development repository. I have fixed it now. 

Also the channels parameter defines the output channel dimension for each layer for all encoders. The number of layers in each encoder will match the length of this list, but the number of layers does not have to match the number of encoders as we can see with the toy dataset experiments.

Hope that helps! I appreciate your feedback.

**Comment by webstah** [comment]
One last note: page 10 of our [preprint](https://arxiv.org/pdf/2309.07138.pdf) has a good visual overview of the model architecture specifications which should help with understanding the code and configurations we used.

**Comment by almas2019** [comment]
Thanks Matthew, following up,  I was wondering how a ~12000 sized input feature is collapsed into a 64 sized vector? Asking this question because the first encoder has incoming channel size = 64. Shouldn't it be 12000? Why is input length is not used in the encoder section? Maybe I am misunderstood something, so clarity is appreciated

**Comment by webstah** [comment]
Also, the way we reduce the length dimension is with a max pooling layer for each convolutional block in the encoder. Each block divides the length by two according to our configuration for the max pooling layers.

**Comment by almas2019** [comment]
Hi Matthew, thank you for your reply.  Mainly my questions relate to trying this method with hypothetically other types of mixed signals, I wanted to clarify inputs and outputs. From what I understand the inputs depend on the yaml file and data loader. I'm not sure where you would get the output for the encoder if you do not want images. There is some missing comments for this in the code.

**Comment by webstah** [comment]
For typical autoencoder architectures, the output dimensions will match the input dimensions and our work is no exception in this regard. We have two fully convolutional architectures available for the BSS task which are capable of taking inputs of variable size (the benefit of using a fully convolutional architecture), so you will not have to explicitly define the input and output dimensions, but you will need to be careful about their shape. Recall from the paper that we use a max pooling layer for each convolutional block with a kernel size of 2 and a stride of 2, effectively dividing the length (or height and width for images) by 2 for each block. This means your input/output features should be divisible by 2 to the power of the number of convolutional blocks in the encoder which is the same amount of blocks in the decoder where we reverse the max pooling with upsampling by a factor of two.
